In [1]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt

C:\Users\Ut Luom\anaconda3\lib\importlib\_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject
  return f(*args, **kwds)


In [17]:
from Clean import outlier as ot

In [9]:
df = pd.read_csv("https://raw.githubusercontent.com/Yimeng-Zhang/feature-engineering-and-feature-selection/master/data/titanic.csv")
data = df[['Pclass', 'Sex', 'Age', 'Fare', 'SibSp', 'Survived']]
data.head()
print(data.shape)

(891, 6)


# Detect random

In [10]:
index, para = ot.outlier_detect_arbitrary(data=data, col='Fare', uper_fence=300, lower_fence=5)
print('Upper bound:', para[0], '\nLower bound:', para[1])

Số lượng outlier: 19
Tỉ lệ outlier được phát hiện: 0.02132435465768799
Upper bound: 300 
Lower bound: 5


In [11]:
# kiểm tra 19 giá trị outlier đó
data.loc[index, 'Fare'].sort_values()

179      0.0000
806      0.0000
732      0.0000
674      0.0000
633      0.0000
597      0.0000
815      0.0000
466      0.0000
481      0.0000
302      0.0000
277      0.0000
271      0.0000
263      0.0000
413      0.0000
822      0.0000
378      4.0125
679    512.3292
737    512.3292
258    512.3292
Name: Fare, dtype: float64

# IQR method

In [20]:
index, para = ot.outlier_detect_IQR(data=data, col='Fare', threshold=0)
print('Upper bound:', para[0], '\nLower bound:', para[1])

Số lượng outlier: 445
Tỉ lệ outlier được phát hiện: 0.4994388327721661
Upper bound: 31.0 
Lower bound: 7.9104


In [21]:
data.loc[index, 'Fare'].sort_values().head()

466    0.0
263    0.0
271    0.0
277    0.0
302    0.0
Name: Fare, dtype: float64

# Mean and Standard Deviation method

In [23]:
index, para = ot.outlier_detect_mean_std(data=data, col='Fare', threshold=3)
print('Upper bound:',para[0], '\nLower bound:', para[1])

Số lượng outlier: 20
Tỉ lệ outlier được phát hiện: 0.02244668911335578
Upper bound: 181.2844937601173 
Lower bound: -116.87607782296811


In [24]:
data.loc[index, 'Fare'].sort_values()

779    211.3375
730    211.3375
689    211.3375
377    211.5000
527    221.7792
716    227.5250
700    227.5250
380    227.5250
557    227.5250
118    247.5208
299    247.5208
311    262.3750
742    262.3750
27     263.0000
341    263.0000
88     263.0000
438    263.0000
258    512.3292
737    512.3292
679    512.3292
Name: Fare, dtype: float64

# MAD: Median and Median Absolute Deviation method

In [26]:
index, para = ot.outlier_detect_MAD(data=data, col='Fare', threshold=3.5)

Số lượng outlier: 160
Tỉ lệ outlier được phát hiện: 0.17957351290684623


ValueError: too many values to unpack (expected 2)

In [27]:
data[255:275]

Pclass     Sex   Age      Fare  SibSp  Survived
255       3  female  29.0   15.2458      0         1
256       1  female   NaN   79.2000      0         1
257       1  female  30.0   86.5000      0         1
258       1  female  35.0  512.3292      0         1
259       2  female  50.0   26.0000      0         1
260       3    male   NaN    7.7500      0         0
261       3    male   3.0   31.3875      4         1
262       1    male  52.0   79.6500      1         0
263       1    male  40.0    0.0000      0         0
264       3  female   NaN    7.7500      0         0
265       2    male  36.0   10.5000      0         0
266       3    male  16.0   39.6875      4         0
267       3    male  25.0    7.7750      1         1
268       1  female  58.0  153.4625      0         1
269       1  female  35.0  135.6333      0         1
270       1    male   NaN   31.0000      0         0
271       3    male  25.0    0.0000      0         1
272       2  female  41.0   19.5000      0         1
273       1    male  37.0   29.7000      0         0
274       3  female   NaN    7.7500      0         1

258, 263, 271 là các giá trị outlier

# Impute with radom value

In [31]:
index, para = ot.outlier_detect_arbitrary(data=data, col='Fare', uper_fence=300, lower_fence=5)
print('Upper bound:', para[0], '\nLower bound:', para[1])

Số lượng outlier: 19
Tỉ lệ outlier được phát hiện: 0.02132435465768799
Upper bound: 300 
Lower bound: 5


In [32]:
data[255:275]

Pclass     Sex   Age      Fare  SibSp  Survived
255       3  female  29.0   15.2458      0         1
256       1  female   NaN   79.2000      0         1
257       1  female  30.0   86.5000      0         1
258       1  female  35.0  512.3292      0         1
259       2  female  50.0   26.0000      0         1
260       3    male   NaN    7.7500      0         0
261       3    male   3.0   31.3875      4         1
262       1    male  52.0   79.6500      1         0
263       1    male  40.0    0.0000      0         0
264       3  female   NaN    7.7500      0         0
265       2    male  36.0   10.5000      0         0
266       3    male  16.0   39.6875      4         0
267       3    male  25.0    7.7750      1         1
268       1  female  58.0  153.4625      0         1
269       1  female  35.0  135.6333      0         1
270       1    male   NaN   31.0000      0         0
271       3    male  25.0    0.0000      0         1
272       2  female  41.0   19.5000      0         1
273       1    male  37.0   29.7000      0         0
274       3  female   NaN    7.7500      0         1

In [33]:
data2 = ot.impute_outlier_with_arbitrary(data=data, outlier_index=index, value=-999, col=['Fare'])
data2[255:275]

Pclass     Sex   Age      Fare  SibSp  Survived
255       3  female  29.0   15.2458      0         1
256       1  female   NaN   79.2000      0         1
257       1  female  30.0   86.5000      0         1
258       1  female  35.0 -999.0000      0         1
259       2  female  50.0   26.0000      0         1
260       3    male   NaN    7.7500      0         0
261       3    male   3.0   31.3875      4         1
262       1    male  52.0   79.6500      1         0
263       1    male  40.0 -999.0000      0         0
264       3  female   NaN    7.7500      0         0
265       2    male  36.0   10.5000      0         0
266       3    male  16.0   39.6875      4         0
267       3    male  25.0    7.7750      1         1
268       1  female  58.0  153.4625      0         1
269       1  female  35.0  135.6333      0         1
270       1    male   NaN   31.0000      0         0
271       3    male  25.0 -999.0000      0         1
272       2  female  41.0   19.5000      0         1
273       1    male  37.0   29.7000      0         0
274       3  female   NaN    7.7500      0         1

# Drop outlier

In [35]:
index, para = ot.outlier_detect_arbitrary(data, 'Fare', 300, 5)

Số lượng outlier: 19
Tỉ lệ outlier được phát hiện: 0.02132435465768799


In [36]:
data3 =  ot.drop_outlier(data=data, outlier_index=index)

# Impute use Median/Mean/Mode

In [41]:
index, para = ot.outlier_detect_arbitrary(data, 'Fare', 300, 5)

Số lượng outlier: 19
Tỉ lệ outlier được phát hiện: 0.02132435465768799


In [43]:
data4 = ot.impute_outlier_with_avg(data=data, col='Fare', outlier_index=index, strategy='mean')
data4[255:275]

Pclass     Sex   Age        Fare  SibSp  Survived
255       3  female  29.0   15.245800      0         1
256       1  female   NaN   79.200000      0         1
257       1  female  30.0   86.500000      0         1
258       1  female  35.0   32.204208      0         1
259       2  female  50.0   26.000000      0         1
260       3    male   NaN    7.750000      0         0
261       3    male   3.0   31.387500      4         1
262       1    male  52.0   79.650000      1         0
263       1    male  40.0   32.204208      0         0
264       3  female   NaN    7.750000      0         0
265       2    male  36.0   10.500000      0         0
266       3    male  16.0   39.687500      4         0
267       3    male  25.0    7.775000      1         1
268       1  female  58.0  153.462500      0         1
269       1  female  35.0  135.633300      0         1
270       1    male   NaN   31.000000      0         0
271       3    male  25.0   32.204208      0         1
272       2  female  41.0   19.500000      0         1
273       1    male  37.0   29.700000      0         0
274       3  female   NaN    7.750000      0         1